##<b><u>Project Title : Sentiment Analysis : Predicting sentiment of COVID-19 tweets</u></b>

## <b> Problem Description </b>

### This challenge asks you to build a classification model to predict the sentiment of COVID-19 tweets.The tweets have been pulled from Twitter and manual tagging has been done then.

### The names and usernames have been given codes to avoid any privacy concerns.

### You are given the following information:
1. Location
2. Tweet At
3. Original Tweet
4. Label

##**Importing Libraries**

In [4]:
# Importing libraries
import pandas as pd
import numpy as np
from numpy import percentile
import matplotlib.pyplot as plt
import seaborn as sns
import matplotlib
import re # for regular expressions
pd.set_option("display.max_colwidth", 200) 
import string
import nltk # for text manipulation
from nltk.stem.porter import *
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer
from tqdm import tqdm
from gensim.models.doc2vec import LabeledSentence
import gensim
from sklearn.linear_model import LogisticRegression
from scipy import stats 
from sklearn import metrics 
from sklearn.metrics import mean_squared_error,mean_absolute_error, make_scorer,classification_report,confusion_matrix,accuracy_score,roc_auc_score,roc_curve
from sklearn.model_selection import train_test_split,cross_val_score,KFold
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score
from sklearn.naive_bayes import BernoulliNB
from sklearn import svm
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
import xgboost as xgb
import warnings 
warnings.filterwarnings("ignore")

%matplotlib inline

In [5]:
# loading csv file in tweet_data dataframe
tweet_data = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Project/sentiment analysis/Coronavirus Tweets.csv',encoding='latin')

## **Exploratory Data Analysis**

In [6]:
#how many rows and columns in dataset
tweet_data.shape

(41157, 6)

in this dataset total 6 columns and 41157 rows(data points) are present 

In [7]:
# details information about dataset
tweet_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41157 entries, 0 to 41156
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   UserName       41157 non-null  int64 
 1   ScreenName     41157 non-null  int64 
 2   Location       32567 non-null  object
 3   TweetAt        41157 non-null  object
 4   OriginalTweet  41157 non-null  object
 5   Sentiment      41157 non-null  object
dtypes: int64(2), object(4)
memory usage: 1.9+ MB


In [8]:
#first 10 datapoints of dataset.
tweet_data.head(10)

UserName  ...           Sentiment
0      3799  ...             Neutral
1      3800  ...            Positive
2      3801  ...            Positive
3      3802  ...            Positive
4      3803  ...  Extremely Negative
5      3804  ...            Positive
6      3805  ...            Positive
7      3806  ...             Neutral
8      3807  ...            Positive
9      3808  ...            Negative

[10 rows x 6 columns]

In [10]:
# Unique numbers of every column datapoints
tweet_data.nunique()

UserName         41157
ScreenName       41157
Location         12220
TweetAt             30
OriginalTweet    41157
Sentiment            5
dtype: int64

###**Details Features information**

* **UserName**: This is the username(encoded in number) unique for every datapoints
* **ScreenName**: This is the ScreenName(encoded in number) unique for every datapoints
* **Location**: places wherefrom tweets are comming
* **TweetAt**: date of tweets data collected
* **OriginalTweet**: Original tweets are stored (This feature is more important in our analysis)
* **Sentiment**: types of sentiment (Positive, Negative, Neutral, Extremely Positive, Extremely Negative) it is dependent variable.


 UserName and ScreenName features are not useful in our analysis, let's drop them

In [13]:
#dropping UserName and ScreenName features
tweet_data = tweet_data.drop(['UserName', 'ScreenName'], axis=1)

In [14]:
#checking duplicate rows 
tweet_data.duplicated().sum()

0

no duplicate rows in dataset

In [15]:
#checking null value in dataset
tweet_data.isnull().sum()

Location         8590
TweetAt             0
OriginalTweet       0
Sentiment           0
dtype: int64

In [16]:
#calculating null value percentage in location feature
tweet_data['Location'].isna().sum()/len(tweet_data) *100

20.87129771363316

* in Location feature have 8590 null value present that is 20.8% But this feature is not important for our sentiment analysis. Hence we will neglect these null values and no nead to treat null value for our analysis

##**Exploring All Features**

In [17]:
#checking unique value feature-wise
tweet_data.nunique()

Location         12220
TweetAt             30
OriginalTweet    41157
Sentiment            5
dtype: int64

*  ### **TweetAt**

In [19]:
#checking unique value in TweetAt feature
tweet_data['TweetAt'].value_counts()

20-03-2020    3448
19-03-2020    3215
25-03-2020    2979
18-03-2020    2742
21-03-2020    2653
22-03-2020    2114
23-03-2020    2062
17-03-2020    1977
08-04-2020    1881
07-04-2020    1843
06-04-2020    1742
24-03-2020    1480
09-04-2020    1471
13-04-2020    1428
26-03-2020    1277
05-04-2020    1131
10-04-2020    1005
02-04-2020     954
11-04-2020     909
03-04-2020     810
12-04-2020     803
04-04-2020     767
16-03-2020     656
01-04-2020     630
27-03-2020     345
31-03-2020     316
14-04-2020     284
29-03-2020     125
30-03-2020      87
28-03-2020      23
Name: TweetAt, dtype: int64

* all tweets belongs to March and April months of 2020


*  ### **Sentiment**

* this feature is our target feature in our analysis

In [18]:
#checking unique value in Sentiment feature
tweet_data['Sentiment'].value_counts()

Positive              11422
Negative               9917
Neutral                7713
Extremely Positive     6624
Extremely Negative     5481
Name: Sentiment, dtype: int64

* there are five types of Sentiment in this Dataset: 
  Positive,  Negative, Neutral, Extremely Positive, Extremely Negative

*   ### **Exploring Sentiment type-wise**

In [20]:
#extracting Extremely Negative tweets
tweet_data[tweet_data['Sentiment']=='Extremely Negative']

Location  ...           Sentiment
4                       NaN  ...  Extremely Negative
20     southampton soxx xxx  ...  Extremely Negative
27      Everywhere You Are!  ...  Extremely Negative
29           Someplace, USA  ...  Extremely Negative
42                  Houston  ...  Extremely Negative
...                     ...  ...                 ...
41115                   NaN  ...  Extremely Negative
41128  Melbourne, Australia  ...  Extremely Negative
41130                   NaN  ...  Extremely Negative
41133               Austria  ...  Extremely Negative
41153                   NaN  ...  Extremely Negative

[5481 rows x 4 columns]

In [21]:
#details comment of Extremely Negative Tweet
tweet_data[tweet_data['Sentiment']=='Extremely Negative'].loc[41133,'OriginalTweet']

'A Singaporean politician who slammed a woman snapped with seven supermarket trolleys during COVID-19 lockdown has been forced to eat humble pie after it turned out she was donating goods to the poor\r\r\nhttps://t.co/EDBXXtdHxl\r\r\n#covid19 #lockdown'

In [22]:
#extracting Negative tweets
tweet_data[tweet_data['Sentiment']=='Negative']

Location  ... Sentiment
9                  BHAVNAGAR,GUJRAT  ...  Negative
24                 Downstage centre  ...  Negative
26                   Ketchum, Idaho  ...  Negative
28                     New York, NY  ...  Negative
30                              NaN  ...  Negative
...                             ...  ...       ...
41127                           NaN  ...  Negative
41131                  Accra, Ghana  ...  Negative
41147                  Brooklyn, NY  ...  Negative
41149              Toronto, Ontario  ...  Negative
41156  i love you so much || he/him  ...  Negative

[9917 rows x 4 columns]

In [24]:
#details comment of Negative Tweet
tweet_data[tweet_data['Sentiment']=='Negative'].loc[28,'OriginalTweet']

'Do you see malicious price increases in NYC? The NYC Department of Consumer and Worker Protection (DCWP) has set up a page to digitally file a complaint. Click here: https://t.co/oEx6Y8mm2K\r\r\n\r\r\nTo file a complaint (use the word"Overcharge") https://t.co/MdMmoBttOP\r\r\n#COVID19 #CovidNYC'

In [23]:
#extracting Neutral tweets
tweet_data[tweet_data['Sentiment']=='Neutral']

Location  ... Sentiment
0                            London  ...   Neutral
7                           Austria  ...   Neutral
10                   Makati, Manila  ...   Neutral
16                     Saudi Arabia  ...   Neutral
17                  Ontario, Canada  ...   Neutral
...                             ...  ...       ...
41141                         India  ...   Neutral
41143              In burning hell.  ...   Neutral
41144                           NaN  ...   Neutral
41152  Wellington City, New Zealand  ...   Neutral
41155                           NaN  ...   Neutral

[7713 rows x 4 columns]

In [25]:
#details comment of Neutral Tweet
tweet_data[tweet_data['Sentiment']=='Neutral'].loc[10,'OriginalTweet']

"All month there hasn't been crowding in the supermarkets or restaurants, however reducing all the hours and closing the malls means everyone is now using the same entrance and dependent on a single supermarket. #manila #lockdown #covid2019 #Philippines https://t.co/HxWs9LAnF9"

In [26]:
#extracting Positive tweets
tweet_data[tweet_data['Sentiment']=='Positive']

Location  ... Sentiment
1                             UK  ...  Positive
2                      Vagabonds  ...  Positive
3                            NaN  ...  Positive
5      ÜT: 36.319708,-82.363649  ...  Positive
6           35.926541,-78.753267  ...  Positive
...                          ...  ...       ...
41140                   Pakistan  ...  Positive
41145              Manhattan, NY  ...  Positive
41146             Gurgaon, India  ...  Positive
41150                       OHIO  ...  Positive
41154                        NaN  ...  Positive

[11422 rows x 4 columns]

In [27]:
#details comment of Positive Tweet
tweet_data[tweet_data['Sentiment']=='Positive'].loc[1,'OriginalTweet']

'advice Talk to your neighbours family to exchange phone numbers create contact list with phone numbers of neighbours schools employer chemist GP set up online shopping accounts if poss adequate supplies of regular meds but not over order'

In [28]:
#extracting Positive tweets
tweet_data[tweet_data['Sentiment']=='Extremely Positive']

Location  ...           Sentiment
11     Pitt Meadows, BC, Canada   ...  Extremely Positive
12                    Horningsea  ...  Extremely Positive
18                 North America  ...  Extremely Positive
22                           NaN  ...  Extremely Positive
23                           NaN  ...  Extremely Positive
...                          ...  ...                 ...
41137                   LES, NYC  ...  Extremely Positive
41138            Los Angeles, CA  ...  Extremely Positive
41142           Juba south sudan  ...  Extremely Positive
41148                        NaN  ...  Extremely Positive
41151                        NaN  ...  Extremely Positive

[6624 rows x 4 columns]

In [29]:
#details comment of Extremely Positive Tweet
tweet_data[tweet_data['Sentiment']=='Extremely Positive'].loc[41138,'OriginalTweet']

'HMU FOR PRICES!! Got great deals going right now. #weedsmokers #StonerFam #420 #bud #Gas #FullSend  #Nelkboys #coronavirus #StimulusChecks #ListenToYourHeart https://t.co/mDMgkOQ8m3'